src/extensions/default/AutoUpdate/unittests.js

```diff
@@ -0,0 +1,491 @@
+
+        validInstallerFilePaths = {
+            "WIN": [_unittestFilesFolder, "Brackets.test.msi"].join("/"),
+            "OSX": [_unittestFilesFolder, "Brackets.test.dmg"].join("/")
+        },
+        invalidInstallerFilePaths = {
+            "WIN": [_unittestFilesFolder, "Brackets.Release.msi"].join("/"),
+            "OSX": [_unittestFilesFolder, "Brackets.Release.dmg"].join("/")
+        };
+
+    describe("AutoUpdate", function () {
+
+        beforeFirst(function () {
+
+            // Create a new window that will be shared by ALL tests in this spec.
+            appInitDone = false;
+            SpecRunnerUtils.createTestWindowAndRun(this, function (w) {
+                testWindow = w;
+                var brackets = testWindow.brackets,
+                    extensionRequire = brackets.test.ExtensionLoader.getRequireContextForExtension("AutoUpdate");
+                AutoUpdate = extensionRequire("main");
+                if(AutoUpdate) {
+                    AutoUpdate.initTestEnv();
+                    var appSupportDirectory = brackets.app.getApplicationSupportDirectory();
+                    updateDir = appSupportDirectory + '/updateTemp';
+                    appInitDone = true;
+                }
+            });
+        });
+
+        afterLast(function () {
+            SpecRunnerUtils.closeTestWindow();
+            testWindow = null;
+        });
+
+
+        beforeEach(function () {
+            fileCheckCompleted = false;
+            updateBarPresent   = false;
+            updateBarDismissed = false;
+            downloadCompleted = false;
+            installerPath = "";
+            errorString   = "";
+        });
+
+        /**
+         * Does a busy wait for a given number of milliseconds
+         * @param {Number} milliSeconds - number of milliSeconds to wait
+         */
+        function waitForMilliSeconds(milliSeconds) {
+            var flag = false;
+
+            setTimeout(function () {
+                flag = true;
+            }, milliSeconds);
+
+            waitsFor(function () {
+                return flag;
+            }, "This should not fail. Please check the timeout values.",
+                milliSeconds + 10); // We give 10 milliSeconds as grace period
+        }
+
+        /**
+         * Check the presence of a Button in Update Bar Prompt
+         * @param {String} btnId - Button Div Id
+         * @param {boolean} click - Should be true If Button need to be clicked else false
+         */
+        function checkUpdateBarButton(btnId, click) {
+            var doc = $(testWindow.document),
+                btn;
+
+            // Test if the update bar button has been displayed.
+            expect(doc.find(UPDATE_BAR).length).toBe(1);
+            btn = doc.find(btnId);
+            expect(btn.length).toBe(1);
+
+            if (btn && click) {
+                // We will wait for 3 seconds so that the Update Bar is visible to the tester.
+                waitForMilliSeconds(3000);
```

Comment:
@niteskum  I am thinking instead of determining the show state based on a wait period, is it possible to listen to state changes and then act upon that. I am not sure if we have done something like this in the other test cases, but can you check and update if that pattern is followed in any test cases?